feature(performance) - Top5 Events

wmak opened the pull request:
- Allow up to 5 `topEvents` to be passed to the event-stats endpoint to get a top-5 breakdown of events over a timerange

wmak committed:
feature(performance) - Initial hacky commit for top n

- TODO:
  - Zerofilling
  - Limiting results more by filter

wmak committed:
feat(performance) - Making this less hacky

- Properly zerofilling now
- Final result format finalized
- TODO: Limiting results by the top events
	- ie. right now non top-n results are coming back

wmak committed:
ref: Moving extra field getting into the query

wmak committed:
fix - gropuby needs to be [] when its not passed

wmak committed:
ref: Filtering properly and fixing porjects

- Using `IN`s so this doesn't retrn absolutely everything
	- TODO: need to actually test this.
- Updating find_ref and timeseries to work with projects

wmak committed:
ref: Getting this to work with transactions

- TODO: testing more fields

wmak committed:
ref: Adding another test

wmak committed:
fix: reference conditions only wants real columns

- top events needs literally everything

markstory committed:
Merge branch 'master' into feature/top-n-events

markstory committed:
Merge branch 'master' into feature/top-n-events

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -67,15 +67,19 @@ def is_real_column(col):
 
 
-def find_reference_event(reference_event):
+def find_reference_event(reference_event, real=True):
```
This is a nit but `real` is not a great argument name. Is that opposed to a `fake` call?

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -619,6 +630,21 @@ def timeseries_query(selected_columns, query, params, rollup, reference_event=No
+    if top_events:
+        event_fields = top_events[0].fields
+        selected_columns += event_fields
+        top_events = {event.slug: find_reference_event(event, real=False) for event in top_events}
```
It would be nice to do one snuba query to get all the event data, instead of one call for each event.

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -67,15 +67,19 @@ def is_real_column(col):
-    if not column_names:
+    columns = (
+        [field for field in reference_event.fields if is_real_column(field)]
+        if real
```
Why do you skip the check for top events?

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -646,18 +672,42 @@ def timeseries_query(selected_columns, query, params, rollup, reference_event=No
+                        [
+                            row
+                            for row in result["data"]
+                            if all(row[field] == event[field] for field in event.keys())
```
Is it possible to have rows that don't match up to one of the reference events?

evanh commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -571,3 +572,240 @@ def test_invalid_interval(self):
+                },
+                project_id=self.project.id,
+            )
+            self.transaction = self.store_event(transaction_data, project_id=self.project.id)
```
Why do you do this 3 times? Shouldn't you at least overwrite some part of the data so you can tell the transactions apart?

wmak replied:
(note: resolved other convos cause they were with reference events and i removed those changes, Replying to this one cause tests are still using this pattern)
I just wanted to create some counts that weren't 1 everywhere, is that problematic? I could offset each event so that they differ.

evanh replied:
It's just that I don't think these will be recorded as 3 separate transactions. At least I've had that happen to me in the tests if I didn't change some of the data between events.

wmak replied:
So they become separate as long as the event_id is different, I now create a unique one for each item in the test using their indexes in the list:
https://github.com/getsentry/sentry/pull/17987/files#diff-3ab46b8a04b6d001891800a63665c90bR657-R662

evanh commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -571,3 +572,240 @@ def test_invalid_interval(self):
+                    "end": iso_format(self.day_ago + timedelta(hours=1, minutes=59)),
+                    "interval": "1h",
+                    "yAxis": "count()",
+                    "field": ["message", "issue.id"],
```
This should be `"issue"` not `"issue.id"` I think. At least, the user can't send `issue.id`. Maybe the frontend overrides that.

markstory replied:
Users can send issue.id if they know about it. But the UI only exposes `issue`.

evanh reviewed (commented):
Other tests I'd like to see:
`topEvents` + `yAxis != count()`
`topEvents` + multiple yAxis

markstory committed:
Merge branch 'master' into feature/top-n-events

wmak committed:
ref - Adding a bulk_find_reference_events

- This way only one snuba query is made for all the reference events in
  a top event call

evanh commented on src/sentry/snuba/discover.py:
```diff
@@ -69,45 +69,57 @@ def is_real_column(col):
+    for reference_event in reference_events:
+        try:
+            project_slug, event_id = reference_event.slug.split(":")
+            project = Project.objects.get(
```
I would pull this check out of the loop, and do one query with all the project IDs. If any of the projects don't exist, you can fail early.

wmak replied:
Is that a safe way to do it since organization is grabbed from reference_event?

evanh replied:
You could grab the org ID from the first reference event. They can't be from different orgs, right?

wmak replied:
Gonna be pulling this out from the PR, because we're no longer passing reference events from the frontend. But yeah the frontend doesn't have a way to do more than 1 org, but I still think it's potentially dangerous (while very edge case-y)

wmak committed:
ref - Refactoring tests so they're easier to work with

fpacifici commented on src/sentry/snuba/discover.py:
```diff
@@ -69,45 +69,57 @@ def is_real_column(col):
-        start=start,
-        end=end,
+        selected_columns=snuba_filter.selected_columns,
+        filter_keys=snuba_filter.filter_keys,
+        aggregations=snuba_filter.aggregations,
+        start=snuba_filter.start,
+        end=snuba_filter.end,
+        groupby=snuba_filter.groupby,
         dataset=Dataset.Discover,
-        limit=1,
-        referrer="discover.find_reference_event",
+        limit=len(reference_events),
+        referrer=referrer,
     )
```
Why can't you just use eventstore to load the event since you know event id and project id ? eventstore is much more efficient for single event loads.

wmak replied:
Going to be removing these changes around reference event :+1:

wmak committed:
ref: Retrieving top 5 in the backend

- Also adding tests for multiple yaxis, and non count axis
- still a WIP, need to do some major cleanups

wmak committed:
Merge branch 'master' into feature/top-n-events

wmak committed:
Revert "Changes to reference_events"

This reverts commit 503ae19d001baf4ebd3fd1509f8936a915af7c75.
- This also removes any other changes to reference events

wmak commented:
Pretty close to ready, I think results are where I'd like them, and this is still 2 queries (1 query + 1 timeseries) instead of (1 reference event + 1 timeseries) Most of these todos have to do with code quality at this point (open to more feedback, but still a WIP).
TODO:
[] Better orderby in top events
[] docstring on get_issues_mapping
[] Explore moving top_events to its own discover function instead of piggybacking timeseries_query
[] cleanup multi yaxis + top events post processing

wmak commented on src/sentry/api/endpoints/organization_events_stats.py:
```diff
@@ -21,11 +21,25 @@ def get(self, request, organization):
 
         def get_event_stats(query_columns, query, params, rollup, reference_event):
+            top_events = None
+            if "topEvents" in request.GET:
```
Would it be more future proof to make this `int`, and set a max of 5 (400 when top>max)?

markstory replied:
Doing an int now with a max of 5 works for me. We can always grow the limit later.

wmak committed:
ref: issues become unknown when group is deleted

wmak committed:
ref: Move most of top_events into its own function

- Also changed the result format to match Mark's suggestion, is now a
  dictionary again, with the keys being all the group by values
  separated by commas
  - Benefit here is that tests became easier again
- This meant moving issues logic as well since otherwise we wouldn't
  know which value in the key is an issue, and so the user still gets
  the short id

wmak committed:
ref: cleanup

- Some changes that were missed from previous commits

wmak committed:
ref: Move multi yaxis dict comprehension into its own function

wmak commented on src/sentry/api/endpoints/organization_events_stats.py:
```diff
@@ -21,11 +21,25 @@ def get(self, request, organization):
+                    params=params,
+                    orderby=["-{}".format(order) for order in orderby],
+                    limit=5,
+                    reference_event=reference_event,
```
Went with `api.organization-event-stats.find-topn` since with the change of `topEvents` to an int, this could be top4 or top3

markstory replied:
topn works for me.

wmak commented on src/sentry/snuba/discover.py:
```diff
@@ -788,6 +790,90 @@ def timeseries_query(selected_columns, query, params, rollup, reference_event=No
+            organization,
+        )
+
+    translated_groupby = [translated_columns.get(field, field) for field in snuba_filter.groupby]
+    if "project.id" in translated_groupby:
+        translated_groupby.remove("project.id")
+        translated_groupby.append("project")
```
I feel like there's probably a better way to do this, but I'm having trouble seeing one 🤔

evanh replied:
Add "project.id": "project" to the translated_columns? This seem fine.

wmak replied:
ooh I like that more :+1:

wmak commented on src/sentry/api/endpoints/organization_events_stats.py:
```diff
@@ -21,11 +21,25 @@ def get(self, request, organization):
+                    limit=5,
+                    reference_event=reference_event,
+                )
+                top_events["columns"] = original_columns
```
Fixed this by making a `top_events_timeseries` function instead, which avoids having to do any thing like this to pass data between the `query` and `timeseries_query`

markstory committed:
Merge branch 'master' into feature/top-n-events

markstory committed:
Merge branch 'master' into feature/top-n-events

wmak committed:
feat: Making the topEvents param an integer

- This allows 1-5 events for top events

wmak committed:
ref: docstring

markstory commented on src/sentry/api/endpoints/organization_events_stats.py:
```diff
@@ -20,17 +21,27 @@ def get(self, request, organization):
+            return Response(
+                {"detail": "Can only get up to {} top events".format(MAX_TOP_EVENTS)}, status=400
+            )
+        elif limit <= 0:
```
Won't this result in an error in all requests that don't set topEvents as you default the value to 0?

wmak replied:
👀 woops, yup. Will fix :+1:

wmak committed:
fix: Ordering by orderby not yaxis

- Also checking top_events so that we don't do the int conversion etc.
  for normal timeseries

wmak committed:
fix: commited WIP test by mistake

evanh reviewed (approved):
This all seems solid to me. Made one comment.

wmak commented on src/sentry/snuba/discover.py:
```diff
@@ -788,6 +790,108 @@ def timeseries_query(selected_columns, query, params, rollup, reference_event=No
+        if field in ["project", "project.id"]:
+            continue
+        values = list({event.get(field) for event in top_events["data"] if field in event})
+        if values and all(values):
```
Lemme add a test for that :+1: sounds like something that might break

wmak replied:
Really good catch :+1: Wouldn't have added a condition for booleans or 0s

markstory reviewed (approved):
Looks good to me, I have a working UI built on top of this.

markstory commented on src/sentry/api/endpoints/organization_events_stats.py:
```diff
@@ -20,7 +21,35 @@ def get(self, request, organization):
+                return discover.top_events_timeseries(
+                    timeseries_columns=query_columns,
+                    selected_columns=request.GET.getlist("field")[:],
+                    user_query=query,
```
Why use a different name than the other methods in this module? To the external consumer this is their 'query'.

wmak replied:
if I use `query` I can't use the `query` function anymore.

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -743,7 +744,9 @@ def key_transaction_timeseries_query(selected_columns, query, params, rollup, re
 
-def timeseries_query(selected_columns, query, params, rollup, reference_event=None, referrer=None):
+def timeseries_query(
+    selected_columns, query, params, rollup, top_events=None, reference_event=None, referrer=None
```
Do you need `top_events` here still?

markstory commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -572,3 +573,373 @@ def test_invalid_interval(self):
+            results = data[
+                ",".join([message, self.event_data[index]["data"]["user"].get("email", "null")])
+            ]
+            assert [{"count": self.event_data[index]["count"] / (3600.0 / 60.0)}] in [
```
Will the value always be `count` when getting other yaxis values?

wmak replied:
Yup, when something is single axis, the values will always be `count`. When its multi axis the results dictionary uses the axis as the key and the values are the same lists of touples but keeps the same format with `count` :+1:

markstory commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -572,3 +573,373 @@ def test_invalid_interval(self):
+                    "start": iso_format(self.day_ago),
+                    "end": iso_format(self.day_ago + timedelta(hours=1, minutes=59)),
+                    "interval": "1h",
+                    "yAxis": ["rpm()", "count()"],
```
This is fannncy :clap:

wmak committed:
ref: extra timeseries param, cleaner way of translated groupby

wmak committed:
fix: Checking for none-ness rather than Falsey

- Certain fields could be 0 or False, this is more a check for None from
  the `.get`

wmak committed:
fix: Fixing timestamps + int

- Couldn't do `timestamp IN []` so used a big `OR` instead.
- Was not handling ints properly, added a cast to string for the `,`
  join

wmak committed:
Merge branch 'master' into feature/top-n-events

wmak committed:
fix: text type means that nothing gets `None` instead of `null`
